src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs

```diff
@@ -2,294 +2,450 @@
+{
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
 
-        private static bool ValidateTxType(Transaction transaction, IReleaseSpec releaseSpec, ref string error)
+        foreach (ITxValidator validator in validators)
         {
-            bool result = transaction.Type switch
+            if (!validator.IsWellFormed(transaction, releaseSpec, out error))
             {
-                TxType.Legacy => true,
-                TxType.AccessList => releaseSpec.UseTxAccessLists,
-                TxType.EIP1559 => releaseSpec.IsEip1559Enabled,
-                TxType.Blob => releaseSpec.IsEip4844Enabled,
-                _ => false
-            };
-
-            if (!result)
-            {
-                error = TxErrorMessages.InvalidTxType(releaseSpec.Name);
                 return false;
             }
+        }
 
+        return true;
+    }
+}
+
+public sealed class IntrinsicGasTxValidator : ITxValidator
+{
+    public static readonly IntrinsicGasTxValidator Instance = new();
+    private IntrinsicGasTxValidator() { }
+
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
+
+        // This is unnecessarily calculated twice - at validation and execution times.
+        var intrinsicGas = IntrinsicGasCalculator.Calculate(transaction, releaseSpec);
+        if (transaction.GasLimit < intrinsicGas)
+        {
+            error = TxErrorMessages.IntrinsicGasTooLow;
+            return false;
+        }
+
+        return true;
+    }
+}
+
+public sealed class ReleaseSpecTxValidator(Func<IReleaseSpec, bool> validate) : ITxValidator
+{
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
+
+        if (!validate(releaseSpec))
+        {
+            error = TxErrorMessages.InvalidTxType(releaseSpec.Name);
+            return false;
+        }
+
+        return true;
+    }
+}
+
+public sealed class ExpectedChainIdTxValidator(ulong chainId) : ITxValidator
+{
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
+
+        if (transaction.ChainId != chainId)
+        {
+            error = TxErrorMessages.InvalidTxChainId(chainId, transaction.ChainId);
+            return false;
+        }
+
+        return true;
+    }
+}
+
+public sealed class GasFieldsTxValidator : ITxValidator
+{
+    public static readonly GasFieldsTxValidator Instance = new();
+    private GasFieldsTxValidator() { }
+
+    public bool IsWellFormed(Transaction transaction, IReleaseSpec releaseSpec, [NotNullWhen(false)] out string? error)
+    {
+        error = null;
+
+        if (!releaseSpec.IsEip1559Enabled)
```

Comment:
I wonder if this is now needed if we guard that in `ReleaseSpecTxValidator`?

Reply:
As mentioned, we can't guard for this without changing the existing behavior.

Reply:
Resolved in the related comment.